File: backend/app/schemas/sale.py
from pydantic import BaseModel, Field
from datetime import datetime, date
from typing import Optional


class SaleBase(BaseModel):
    """Base schema for Sale"""
    product_id: int
    customer_name: str = Field(..., min_length=1, max_length=255)
    customer_cpf_cnpj: Optional[str] = Field(None, max_length=14)
    customer_email: Optional[str] = Field(None, max_length=255)
    customer_phone: Optional[str] = Field(None, max_length=20)

    # Endereço do Cliente
    customer_cep: Optional[str] = Field(None, max_length=8)
    customer_logradouro: Optional[str] = Field(None, max_length=255)
    customer_numero: Optional[str] = Field(None, max_length=20)
    customer_complemento: Optional[str] = Field(None, max_length=100)
    customer_bairro: Optional[str] = Field(None, max_length=100)
    customer_cidade: Optional[str] = Field(None, max_length=100)
    customer_uf: Optional[str] = Field(None, max_length=2)
    customer_codigo_municipio: Optional[str] = Field(None, max_length=7)

    quantity: int = Field(..., gt=0)
    unit_price: float = Field(..., ge=0)
    total_value: float = Field(..., ge=0)
    sale_date: date

    # Campos fiscais opcionais
    natureza_operacao: Optional[str] = Field(None, max_length=100)
    cfop: Optional[str] = Field(None, max_length=4)
    origin_channel: Optional[str] = Field(None, max_length=50)
    origin_order_id: Optional[str] = Field(None, max_length=100)
    notes: Optional[str] = None


class SaleCreate(SaleBase):
    """Schema for creating a new Sale"""
    pass


class SaleUpdate(BaseModel):
    """Schema for updating a Sale"""
    product_id: Optional[int] = None
    customer_name: Optional[str] = Field(None, min_length=1, max_length=255)
    customer_cpf_cnpj: Optional[str] = Field(None, max_length=14)
    customer_email: Optional[str] = Field(None, max_length=255)
    customer_phone: Optional[str] = Field(None, max_length=20)

    # Endereço do Cliente
    customer_cep: Optional[str] = Field(None, max_length=8)
    customer_logradouro: Optional[str] = Field(None, max_length=255)
    customer_numero: Optional[str] = Field(None, max_length=20)
    customer_complemento: Optional[str] = Field(None, max_length=100)
    customer_bairro: Optional[str] = Field(None, max_length=100)
    customer_cidade: Optional[str] = Field(None, max_length=100)
    customer_uf: Optional[str] = Field(None, max_length=2)
    customer_codigo_municipio: Optional[str] = Field(None, max_length=7)

    quantity: Optional[int] = Field(None, gt=0)
    unit_price: Optional[float] = Field(None, ge=0)
    total_value: Optional[float] = Field(None, ge=0)
    status: Optional[str] = Field(None, pattern="^(pending|completed|cancelled)$")
    sale_date: Optional[date] = None

    # Campos fiscais opcionais
    natureza_operacao: Optional[str] = Field(None, max_length=100)
    cfop: Optional[str] = Field(None, max_length=4)
    origin_channel: Optional[str] = Field(None, max_length=50)
    origin_order_id: Optional[str] = Field(None, max_length=100)
    notes: Optional[str] = None


class SaleResponse(SaleBase):
    """Schema for Sale response"""
    id: int
    workspace_id: int
    status: str

    # Campos NF-e
    nfe_status: Optional[str] = None
    nfe_id_partner: Optional[str] = None
    nfe_chave: Optional[str] = None
    nfe_numero: Optional[int] = None
    nfe_serie: Optional[int] = None
    nfe_xml_url: Optional[str] = None
    nfe_danfe_url: Optional[str] = None
    nfe_protocolo: Optional[str] = None
    nfe_issued_at: Optional[datetime] = None
    nfe_rejection_reason: Optional[str] = None
    nfe_rejection_code: Optional[str] = None
    nfe_cancelled_at: Optional[datetime] = None
    nfe_cancellation_reason: Optional[str] = None

    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: admin/src/lib/validations/sale.ts
import { z } from 'zod';

export const saleSchema = z.object({
  product_id: z.number({ message: 'Produto é obrigatório' })
    .positive('Selecione um produto válido'),

  customer_name: z.string({ message: 'Nome do cliente é obrigatório' })
    .min(1, 'Nome do cliente não pode estar vazio')
    .max(255, 'Nome muito longo'),

  customer_cpf_cnpj: z.string().max(14).optional(),
  customer_email: z.string().email('Email inválido').max(255).optional().or(z.literal('')),
  customer_phone: z.string().max(20).optional(),

  // Endereço do Cliente
  customer_cep: z.string().max(8).optional(),
  customer_logradouro: z.string().max(255).optional(),
  customer_numero: z.string().max(20).optional(),
  customer_complemento: z.string().max(100).optional(),
  customer_bairro: z.string().max(100).optional(),
  customer_cidade: z.string().max(100).optional(),
  customer_uf: z.string().max(2).optional(),
  customer_codigo_municipio: z.string().max(7).optional(),

  quantity: z.number({ message: 'Quantidade é obrigatória' })
    .int('Quantidade deve ser um número inteiro')
    .positive('Quantidade deve ser maior que zero'),

  unit_price: z.number({ message: 'Preço unitário é obrigatório' })
    .positive('Preço unitário deve ser maior que zero'),

  total_value: z.number({ message: 'Valor total é obrigatório' })
    .positive('Valor total deve ser maior que zero'),

  sale_date: z.string({ message: 'Data da venda é obrigatória' })
    .min(1, 'Data da venda é obrigatória'),

  status: z.enum(['pending', 'completed', 'cancelled']).optional(),

  // Campos fiscais
  natureza_operacao: z.string().max(100).optional(),
  cfop: z.string().max(4).optional(),
  origin_channel: z.string().max(50).optional(),
  origin_order_id: z.string().max(100).optional(),
  notes: z.string().optional(),
}).refine(
  (data) => {
    // Validar se total_value = quantity * unit_price (com margem de erro para arredondamento)
    const calculated = data.quantity * data.unit_price;
    const diff = Math.abs(calculated - data.total_value);
    return diff < 0.01; // Margem de 1 centavo
  },
  {
    message: 'Valor total deve ser igual a quantidade × preço unitário',
    path: ['total_value'],
  }
);

export type SaleFormData = z.infer<typeof saleSchema>;

export const saleUpdateSchema = saleSchema.partial();

export type SaleUpdateFormData = z.infer<typeof saleUpdateSchema>;
